[WIP] Implement empty-tensor operators for XPU backend

fengyuan14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119157
* #117234
* __->__ #118073
* #117970

Features: https://github.com/pytorch/pytorch/issues/114835

1. Enhance building capability for `libtorch_xpu.so` with `sycl_add_library`. `libtorch_xpu.so` is XPU backend implementation of PyTorch ATen tensor library, including
   kernel part in SYCL and host part in C++. Enable separate compilation by distinguishing SYCL and C++ sources and using SYCL and C++ compilers compile them separately.
2. Add two Tensor factory ATen operators for XPU backend.

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

arthuryuan1987 commented:
Convert to Draft since the PR depends on [Intel GPU Runtime Upstreaming for Device](https://github.com/pytorch/pytorch/pull/116869)

fengyuan14 committed:
Implement empty-tensor operators for XPU backend

Features: https://github.com/pytorch/pytorch/issues/114835

1. Enhance building capability for `libtorch_xpu.so` with `sycl_add_library`. `libtorch_xpu.so` is XPU backend implementation of PyTorch ATen tensor library, including
   kernel part in SYCL and host part in C++. Enable separate compilation by distinguishing SYCL and C++ sources and using SYCL and C++ compilers compile them separately.
2. Add two Tensor factory ATen operators for XPU backend.

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

[ghstack-poisoned]

jgong5 commented on test/test_xpu.py:
```diff
@@ -0,0 +1,55 @@
+# not torch.backends.xpu.is_avaliable
+@unittest.skipIf(IS_WINDOWS, "XPU operators testing is disabled")
+class TestXPUOperators(TestCase):
+    def test_tensor_factory(self):
```
Suggest not to pack multiple tests into a single UT. Separate them and name them explicity with `test_empty` and `test_empty_strided`.

jgong5 commented on aten/src/ATen/xpu/EmptyTensor.h:
```diff
@@ -0,0 +1,42 @@
+
+namespace at::detail {
+
+// XXX: add TORCH_XPU_API
```
Do we need to expose them as API?

arthuryuan1987 replied:
Yes, here we will follow CUDA exposure.

jgong5 commented on aten/src/ATen/native/xpu/TensorFactories.cpp:
```diff
@@ -0,0 +1,35 @@
+Tensor empty_xpu(IntArrayRef size, c10::optional<ScalarType> dtype_opt, c10::optional<Layout> layout_opt, c10::optional<Device> device_opt, c10::optional<bool> pin_memory_opt, c10::optional<c10::MemoryFormat> memory_format_opt) {
+  Tensor result = at::detail::empty_xpu(size, dtype_opt, layout_opt, device_opt, pin_memory_opt, memory_format_opt);
+  // See Note [Enabling Deterministic Operations]
+  TORCH_CHECK(!(C10_UNLIKELY(at::globalContext().deterministicAlgorithms() && at::globalContext().deterministicFillUninitializedMemory())), "XPU backend doesn't support deterministic implementation for empty ...")
```
May I know why `empty` call is related to determinism? It doesn't initialize memory, right?

jgong5 replied:
Hmm, I checked cuda implementation `empty_cuda` and didn't find the checking of deterministic algorithms. Do I miss anything?

arthuryuan1987 replied:
Yes, I will retrieve it back, if we have fill kernel.

jgong5 commented on aten/src/ATen/xpu/EmptyTensor.cpp:
```diff
@@ -0,0 +1,96 @@
+    c10::optional<c10::MemoryFormat> memory_format_opt) {
+  const auto device = device_or_default(device_opt);
+  TORCH_INTERNAL_ASSERT(device.is_xpu());
+  // XXX
+  // const c10::DeviceGuard device_guard(device);
+  auto* allocator = at::getCPUAllocator();
```
Hack... Are you going to fix it in this PR?

arthuryuan1987 replied:
Yes. It depends on runtime PRs landed.

jgong5 commented on aten/src/ATen/xpu/EmptyTensor.cpp:
```diff
@@ -0,0 +1,96 @@
+    c10::optional<Device> device_opt) {
+  const auto device = device_or_default(device_opt);
+  TORCH_INTERNAL_ASSERT(device.is_xpu());
+  // XXX
+  // const c10::DeviceGuard device_guard(device);
+  auto* allocator = at::getCPUAllocator();
```
ditto

fengyuan14 committed:
Update on "[WIP] Implement empty-tensor operators for XPU backend"


Features: https://github.com/pytorch/pytorch/issues/114835

1. Enhance building capability for `libtorch_xpu.so` with `sycl_add_library`. `libtorch_xpu.so` is XPU backend implementation of PyTorch ATen tensor library, including
   kernel part in SYCL and host part in C++. Enable separate compilation by distinguishing SYCL and C++ sources and using SYCL and C++ compilers compile them separately.
2. Add two Tensor factory ATen operators for XPU backend.

Signed-off-by: Feng Yuan <feng1.yuanintel.com>

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

fengyuan14 committed:
Update on "[WIP] Implement empty-tensor operators for XPU backend"


Features: https://github.com/pytorch/pytorch/issues/114835

1. Enhance building capability for `libtorch_xpu.so` with `sycl_add_library`. `libtorch_xpu.so` is XPU backend implementation of PyTorch ATen tensor library, including
   kernel part in SYCL and host part in C++. Enable separate compilation by distinguishing SYCL and C++ sources and using SYCL and C++ compilers compile them separately.
2. Add two Tensor factory ATen operators for XPU backend.

Signed-off-by: Feng Yuan <feng1.yuanintel.com>

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

fengyuan14 committed:
Update on "[WIP] Implement empty-tensor operators for XPU backend"


Features: https://github.com/pytorch/pytorch/issues/114835

1. Enhance building capability for `libtorch_xpu.so` with `sycl_add_library`. `libtorch_xpu.so` is XPU backend implementation of PyTorch ATen tensor library, including
   kernel part in SYCL and host part in C++. Enable separate compilation by distinguishing SYCL and C++ sources and using SYCL and C++ compilers compile them separately.
2. Add two Tensor factory ATen operators for XPU backend.

Signed-off-by: Feng Yuan <feng1.yuanintel.com>

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

fengyuan14 committed:
Update on "[WIP] Implement empty-tensor operators for XPU backend"


Features: https://github.com/pytorch/pytorch/issues/114835

1. Enhance building capability for `libtorch_xpu.so` with `sycl_add_library`. `libtorch_xpu.so` is XPU backend implementation of PyTorch ATen tensor library, including
   kernel part in SYCL and host part in C++. Enable separate compilation by distinguishing SYCL and C++ sources and using SYCL and C++ compilers compile them separately.
2. Add two Tensor factory ATen operators for XPU backend.

Signed-off-by: Feng Yuan <feng1.yuanintel.com>

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

fengyuan14 committed:
Update on "[WIP] Implement empty-tensor operators for XPU backend"


Features: https://github.com/pytorch/pytorch/issues/114835

1. Enhance building capability for `libtorch_xpu.so` with `sycl_add_library`. `libtorch_xpu.so` is XPU backend implementation of PyTorch ATen tensor library, including
   kernel part in SYCL and host part in C++. Enable separate compilation by distinguishing SYCL and C++ sources and using SYCL and C++ compilers compile them separately.
2. Add two Tensor factory ATen operators for XPU backend.

Signed-off-by: Feng Yuan <feng1.yuanintel.com>

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

fengyuan14 committed:
Update on "[WIP] Implement empty-tensor operators for XPU backend"


Features: https://github.com/pytorch/pytorch/issues/114835

1. Enhance building capability for `libtorch_xpu.so` with `sycl_add_library`. `libtorch_xpu.so` is XPU backend implementation of PyTorch ATen tensor library, including
   kernel part in SYCL and host part in C++. Enable separate compilation by distinguishing SYCL and C++ sources and using SYCL and C++ compilers compile them separately.
2. Add two Tensor factory ATen operators for XPU backend.

Signed-off-by: Feng Yuan <feng1.yuanintel.com>

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

fengyuan14 committed:
Update on "[WIP] Implement empty-tensor operators for XPU backend"


Features: https://github.com/pytorch/pytorch/issues/114835

1. Enhance building capability for `libtorch_xpu.so` with `sycl_add_library`. `libtorch_xpu.so` is XPU backend implementation of PyTorch ATen tensor library, including
   kernel part in SYCL and host part in C++. Enable separate compilation by distinguishing SYCL and C++ sources and using SYCL and C++ compilers compile them separately.
2. Add two Tensor factory ATen operators for XPU backend.

Signed-off-by: Feng Yuan <feng1.yuanintel.com>

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]
